cluster-autoscaler/cloudprovider/openstack/openstack_nodegroup.go

```diff
@@ -0,0 +1,288 @@
+		return fmt.Errorf("size increase must be positive")
+	}
+
+	size, err := ng.openstackManager.NodeGroupSize(ng.id)
+	if err != nil {
+		return fmt.Errorf("could not check current nodegroup size: %v", err)
+	}
+	if size+delta > ng.MaxSize() {
+		return fmt.Errorf("size increase too large, desired:%d max:%d", size+delta, ng.MaxSize())
+	}
+
+	updatePossible, currentStatus, err := ng.openstackManager.CanUpdate()
+	if err != nil {
+		return fmt.Errorf("can not increase node count: %v", err)
+	}
+	if !updatePossible {
+		return fmt.Errorf("can not add nodes, cluster is in %s status", currentStatus)
+	}
+	klog.V(0).Infof("Increasing size by %d, %d->%d", delta, *ng.targetSize, *ng.targetSize+delta)
+	*ng.targetSize += delta
+
+	err = ng.openstackManager.UpdateNodeCount(ng.id, *ng.targetSize)
+	if err != nil {
+		return fmt.Errorf("could not increase cluster size: %v", err)
+	}
+
+	// Block until cluster has gone into update status and then completed
+	err = ng.WaitForClusterStatus(clusterStatusUpdateInProgress, waitForUpdateStatusTimeout)
+	if err != nil {
+		return fmt.Errorf("scale up error: %v", err)
+	}
+	err = ng.WaitForClusterStatus(clusterStatusUpdateComplete, waitForCompleteStatusTimout)
+	if err != nil {
+		return fmt.Errorf("scale up error: %v", err)
+	}
+	return nil
+}
+
+// DeleteNodes deletes a set of nodes chosen by the autoscaler.
+//
+// The process of deletion depends on the implementation of OpenstackManager,
+// but this function handles what should be common between all implementations:
+//   - simultaneous but separate calls from the autoscaler are batched together
+//   - does not allow scaling while the cluster is already in an UPDATE_IN_PROGRESS state
+//   - after scaling down, blocks until the cluster has reached UPDATE_COMPLETE
+func (ng *OpenstackNodeGroup) DeleteNodes(nodes []*apiv1.Node) error {
+
+	// Batch simultaneous deletes on individual nodes
+	ng.nodesToDeleteMutex.Lock()
+	// Check that these nodes would not make the batch delete more nodes than the minimum would allow
+	if *ng.targetSize-len(ng.nodesToDelete)-len(nodes) < ng.MinSize() {
+		ng.nodesToDeleteMutex.Unlock()
+		return fmt.Errorf("deleting nodes would take nodegroup below minimum size")
+	}
+	ng.nodesToDelete = append(ng.nodesToDelete, nodes...)
+	ng.nodesToDeleteMutex.Unlock()
+
+	ng.clusterUpdateMutex.Lock()
+	defer ng.clusterUpdateMutex.Unlock()
+
+	ng.nodesToDeleteMutex.Lock()
+	if len(ng.nodesToDelete) == 0 {
+		// Deletion was handled by another goroutine
+		ng.nodesToDeleteMutex.Unlock()
+		return nil
+	}
+	ng.nodesToDeleteMutex.Unlock()
+
+	// This goroutine has the clusterUpdateMutex, so will be the one
+	// to actually delete the nodes. While this goroutine waits, others
+	// will add their nodes to nodesToDelete and block at acquiring
+	// the clusterUpdateMutex lock. One they get it, the deletion will
+	// already be done and they will return above at the check
+	// for len(ng.nodesToDelete) == 0.
+	time.Sleep(deleteNodesBatchingDelay)
```

Comment:
I like that this saves on Heat API hits, but for this function I think it might be better to just issue the updates as they come in and not poll the API waiting for updates to finish.  I think the ng.nodesToDelete should be appended as deletes come in (stored across calls) and just resubmitted.  Once a scale up is issued, the ng.nodesToDelete can be purged to prevent accidental removal of a node with a reused IP.  There should be no harm if 100s of IPs are specified when removing a single node as long as we know there were no scale ups operations in between scale down.